### DISCUSSION TITLE
How is long text handled?

### DISCUSSION THREAD

**Discussion by YanDavKMS** [michaelfeil/infinity#89] [Q&A] [answered]
Hey,

I'm trying to understand, what happens if we send a long text, which is longer from the model max length?
Will it be truncated by the tokenizer?
If not, what happens if the model gets a longer text than it's max length?

Thanks.

**Comment by michaelfeil** [comment, marked as answer]
Hi @YanDavKMS it will be truncated, same behaviour as `sentence-transformers`. Longest is truncated first, for `bge-large-v1.5` this would result only in the first 512 tokens to be embedded.

**Reply by YanDavKMS** [reply]
Thank you for the answer.

I have sent a text longer than 512 and 2 things happened that are not clear to me:

1. I get the following warning "Token indices sequence length is longer than the specified maximum sequence length for this model (691 > 512). Running this sequence through the model will result in indexing errors" so if it is truncated why do I still get this error?
2. The returned tokens count from the embedding endpoint in infinity is 691

**Reply by michaelfeil** [reply]
The sententence gets tokenized twice. Once for counting the tokens of the sentence correctly (691). If you use the debugger, you will notice that the actual tokens / pytorch input_ids in the forward pass is 512.

**Reply by YanDavKMS** [reply]
Thank you very much for the explanation.

Last point. from the usage perspective perhaps it is better to return (also) the real tokens usage.

**Reply by michaelfeil** [reply]
What's the rationale behind that, I am curious?

**Reply by YanDavKMS** [reply]
There are a few reasons why I would like to know how many tokens did I really run the model on

- For statistics and performance measurements
- For charging purposes, same as if I would use OpenAI embeddings, I get the real tokens usage
- It would make it clearer that even though I sent a longer text I actually got embedding on truncated text

**Reply by michaelfeil** [reply]
Thanks for the insights. I'll slowly pivot in this direction.
